core/amber/src/main/scala/edu/uci/ics/amber/engine/common/promise/PromiseManager.scala

```diff
@@ -0,0 +1,234 @@
+          // execute this one
+          registerSyncPromise(ctx, ctx)
+          invokePromise(ctx, call)
+        } else {
+          // otherwise, enqueue it
+          queuedInvocations.enqueue(payload)
+        }
+
+      // handle synchronous promise created by other promise
+      case PromiseInvocation(
+            ctx: ChildPromiseContext,
+            call: PromiseBody[_] with SynchronizedInvocation
+          ) =>
+        if (syncPromiseRoot == null || ctx.root == syncPromiseRoot) {
+          // if there is no other executing synchronous promise,
+          // or this promise has same root context (chain of promises),
+          // execute this one.
+          registerSyncPromise(ctx.root, ctx)
+          invokePromise(ctx, call)
+        } else {
+          // otherwise, enqueue it
+          queuedInvocations.enqueue(payload)
+        }
+
+      // trivial case
+      case p: PromiseInvocation =>
+        // set context
+        promiseContext = p.context
+        try {
+          // execute it
+          promiseHandler(p.call)
+        } catch {
+          case e: Throwable =>
+            // if error occurs, return it to the sender.
+            returning(e)
+        }
+    }
+
+    // execute queued sync promises
+    tryInvokeNextSyncPromise()
+  }
+
+  // send a control message to another actor, and keep the handle.
+  def schedule[T](cmd: PromiseBody[T], on: ActorVirtualIdentity): Promise[T] = {
+    val (promise, ctx) = createPromise[T]()
+    controlOutputPort.sendTo(on, PromiseInvocation(ctx, cmd))
+    promise
+  }
+
+  // send a grouped control message to other actors, and keep the handle.
+  def schedule[T: ClassTag](seq: (PromiseBody[T], ActorVirtualIdentity)*): Promise[Seq[T]] = {
+    val promise = WorkflowPromise[Seq[T]](promiseContext)
+    if (seq.isEmpty) {
+      // if the sequence is empty, resolve the promise immediately
+      promise.setValue(Seq.empty)
+    } else {
+      unCompletedGroupPromises.add(
+        GroupedWorkflowPromise[T](promiseID, promiseID + seq.length, promise)
+      )
+      seq.foreach {
+        case (body, virtualIdentity) =>
+          val ctx = mkPromiseContext()
+          promiseID += 1
+          controlOutputPort.sendTo(virtualIdentity, PromiseInvocation(ctx, body))
+      }
+    }
+    promise
+  }
+
+  def returning(value: Any): Unit = {
+    // returning should be used at most once per context
+    if (promiseContext != null) {
+      controlOutputPort.sendTo(promiseContext.sender, ReturnPayload(promiseContext, value))
+      exitCurrentPromise()
+    }
+  }
+
+  def returning(): Unit = {
+    // returning should be used at most once per context
+    if (promiseContext != null) {
+      controlOutputPort.sendTo(
+        promiseContext.sender,
+        ReturnPayload(promiseContext, PromiseCompleted())
+      )
+      exitCurrentPromise()
+    }
+  }
+
+  @inline
+  def createPromise[T](): (WorkflowPromise[T], PromiseContext) = {
```

Comment:
If I understand correctly, createPromise() creates a context that is sent with the ControlPayload to the next actor. So, it doesn't need to return WorkflowPromise. It should just return PromiseContext that will be sent to the next actor (`ctx`).

Reply:
Lets break this function down into 2 parts - createPromiseContext() and registerCallerPromiseContext

Reply:
Actually I am thinking there is no need for keeping a `CallerPromiseContext` as a data member. If A calls C, it passes its context to C, which C should register in map, and then call handler. At the same time, if B calls C, its `CallerPromiseContext` should also be put in the same map.